# Analysis of the Startups of the New Apps for Books Lovers 
# Contents <a id='back'></a>

1. Introduction
    
    
2. Data Pra-processing
   

3. Exploratory Data Analysis (EDA)


4. Overall Conclusion & Recommendations

## Introduction

The coronavirus took the entire world by surprise, changing everyone's daily routine. City dwellers no longer spent their free time outside, going to cafes and malls; more people were at home, reading books. That attracted the attention of startups that rushed to develop new apps for book lovers. 

**Purposes**

We've been given a database of one of the services competing in this market. It contains data on books, publishers, authors, and customer ratings and reviews of books. This information will be used to generate a value proposition for a new product.

**Note: Program will used is SQL**

## Data Pra-processing <a id='intro'></a>
### Loading Data

In [1]:
# Load the required libraries

import pandas as pd
import numpy as np
from sqlalchemy import create_engine

db_config = {'user': 'practicum_student',         # nama pengguna
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # kata sandi
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # port koneksi
             'db': 'data-analyst-final-project-db'}          # nama database

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

**Description of the data**

We have five tables, as follows:

`books` Contains data on books:

- `book_id`
- `author_id`
- `title`
- `num_pages` — number of pages
- `publication_date`
- `publisher_id`

`authors` Contains data on authors:

- `author_id`
- `author`

`publishers` Contains data on publishers:

- `publisher_id`
- `publisher`

`ratings` Contains data on user ratings:

- `rating_id`
- `book_id`
- `username` — the name of the user who rated the book
- `rating`

`reviews` Contains data on customer reviews:

- `review_id`
- `book_id`
- `username` — the name of the user who reviewed the book
- `text` — the text of the review

### Table of `books`

`books` Contains data on books:

- `book_id`
- `author_id`
- `title`
- `num_pages` — number of pages
- `publication_date`
- `publisher_id`


In [2]:
# See the table of books

pd.io.sql.read_sql(
'''
SELECT
    *
FROM 
    books
''',
con = engine)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

In [3]:
# Check missing values in the table of books

pd.io.sql.read_sql(
'''
SELECT
    *
FROM
    books
WHERE
    books ISNULL
''',
con = engine)

Empty DataFrame
Columns: [book_id, author_id, title, num_pages, publication_date, publisher_id]
Index: []

In [4]:
# Get the first and the last publication date of book

pd.io.sql.read_sql(
'''
SELECT
    MIN (publication_date),
    MAX (publication_date)
FROM
    books
''',
con = engine)

min         max
0  1952-12-01  2020-03-31

### Table of `authors`

`authors` Contains data on authors:

- `author_id`
- `author`


In [5]:
# See the table of authors

pd.io.sql.read_sql(
'''
SELECT
    *
FROM
    authors
''',
con = engine)

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

In [6]:
# Check missing values in the table of authors

pd.io.sql.read_sql(
'''
SELECT
    *
FROM
    authors
WHERE
    authors ISNULL
''',
con = engine)

Empty DataFrame
Columns: [author_id, author]
Index: []

### Table of `publishers`

`publishers` Contains data on publishers:

- `publisher_id`
- `publisher`


In [7]:
# See the table of publishers

pd.io.sql.read_sql(
'''
SELECT 
    *
FROM 
    publishers
''',
con = engine)

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

In [8]:
# Check missing values in the table of publishers

pd.io.sql.read_sql(
'''
SELECT *
FROM publishers
WHERE publishers ISNULL
''',
con = engine)

Empty DataFrame
Columns: [publisher_id, publisher]
Index: []

### Table of `ratings`

`ratings` Contains data on user ratings:

- `rating_id`
- `book_id`
- `username` — the name of the user who rated the book
- `rating`


In [9]:
# See the table of ratings

pd.io.sql.read_sql(
'''
SELECT
    *
FROM
    ratings
''',
con = engine)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

In [10]:
# Check missing values in the table of ratings

pd.io.sql.read_sql(
'''
SELECT *
FROM ratings
WHERE ratings ISNULL
''',
con = engine)

Empty DataFrame
Columns: [rating_id, book_id, username, rating]
Index: []

### Table of `reviews`

`reviews` Contains data on customer reviews:

- `review_id`
- `book_id`
- `username` — the name of the user who reviewed the book
- `text` — the text of the review

In [11]:
# See the table of reviews

pd.io.sql.read_sql(
'''
SELECT
    *
FROM
    reviews
''',
con = engine)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

In [12]:
# Check missing values in the table of reviews

pd.io.sql.read_sql(
'''
SELECT
    *
FROM
    reviews
WHERE
    reviews ISNULL
''',
con = engine)

Empty DataFrame
Columns: [review_id, book_id, username, text]
Index: []

### Initial Conslusions

- We have tables, as follows:
    1. The `books` table that contains data on books such as `book_id`, `author_id`, `title`, `num_pages`, `publication_date` and `publisher_id`.
    2. The `authors` table that contains data on authors such as `author_id` and `author name`.
    3. The `publishers` table that contains data on publishers such as `publisher_id` and `publisher name`.
    4. The `ratings` table that contains data on user ratings such as `rating_id` - refers to the rating order by the user, `book_id`, `username` — the name of the user who rated the book and `rating` - scale 1 to 5.
    5. The `reviews` table that contains data on customer reviews such as `review_id` - refers to the text review order by the user, `book_id`, `username` and `text` — the text of the review.
- There is no missing values.
- The publication date of books are from 1 Dec 1952 to 31 March 2020.

## Exploratory Data Analysis (EDA)

### Find the number of books released after January 1, 2000.

In [13]:
# Calculate the number of books released after January 1, 2000

pd.io.sql.read_sql(
'''
SELECT
    COUNT(title)
FROM 
    books
WHERE 
    publication_date > '2000-01-01'
''', 
con = engine)

count
0    819

**Findings**:

- The total of books released after January 1, 2000 are 819 books.


### Find the number of user reviews and the average rating for each book.

Step by step:
- Using CTE (Common Table Expression), we created two new tables, as follows:
    1. First table to get the number of reviews of each book.
    2. Second table to get the average of ratings of each book.
- These two tables merge using INNER JOIN method on book_id.

In [14]:
# Calculate the number of reviews and the average rating of each book

pd.io.sql.read_sql(
'''
WITH agg_reviews_per_book AS (
    SELECT
        reviews.book_id,
        COUNT(reviews.text) AS cnt_reviews
    FROM
        reviews
    GROUP BY
        reviews.book_id
    ORDER BY
        cnt_reviews DESC
)
, avg_ratings_per_book AS(
    SELECT
        ratings.book_id,
        AVG(ratings.rating) AS avg_ratings
    FROM 
        ratings
    GROUP BY
        ratings.book_id
    ORDER BY
        avg_ratings DESC
)
SELECT 
    agg_reviews_per_book.book_id,
    cnt_reviews,
    avg_ratings
FROM
    agg_reviews_per_book
    INNER JOIN avg_ratings_per_book on agg_reviews_per_book.book_id = avg_ratings_per_book.book_id
''',
con = engine)

book_id  cnt_reviews  avg_ratings
0          1            2     3.666667
1          2            1     2.500000
2          3            3     4.666667
3          4            2     4.500000
4          5            4     4.000000
..       ...          ...          ...
989      996            3     3.666667
990      997            3     3.400000
991      998            4     3.200000
992      999            2     4.500000
993     1000            4     3.833333

[994 rows x 3 columns]

In [15]:
# Calculate the number of reviews and the average rating of each book

pd.io.sql.read_sql(
'''
WITH main AS (
    WITH agg_reviews_per_book AS (
        SELECT
            reviews.book_id,
            COUNT(reviews.text) AS cnt_reviews
        FROM
            reviews
        GROUP BY
            reviews.book_id
        ORDER BY
            cnt_reviews DESC
    )
    , avg_ratings_per_book AS(
        SELECT
            ratings.book_id,
            AVG(ratings.rating) AS avg_ratings
        FROM 
            ratings
        GROUP BY
            ratings.book_id
        ORDER BY
            avg_ratings DESC
    )
    SELECT 
        agg_reviews_per_book.book_id,
        cnt_reviews,
        avg_ratings
    FROM
        agg_reviews_per_book
        INNER JOIN avg_ratings_per_book on agg_reviews_per_book.book_id = avg_ratings_per_book.book_id
)
SELECT
    MIN(cnt_reviews) AS min_reviews_per_book,
    MAX(cnt_reviews) AS max_reviews_per_book,
    MIN(avg_ratings) AS min_ratings_per_book,
    MAX(avg_ratings) AS max_ratings_per_book
FROM
    main
''',
con = engine)

min_reviews_per_book  max_reviews_per_book  min_ratings_per_book  \
0                     1                     7                   1.5   

   max_ratings_per_book  
0                   5.0

**Findings**:

- There are 994 books that receive reviews from the users, with the range of reviews per book being 1 of 7 reviews and the range of average ratings is from a scale of 1.5 to 5.

### Identify the publisher that has released the greatest number of books with more than 50 pages.

In [16]:
# Calculate the publisher that has released the greatest number of books with more than 50 pages

pd.io.sql.read_sql(
'''
SELECT 
    publisher,
    COUNT(publisher) AS cnt
FROM 
    publishers
    LEFT JOIN books ON books.publisher_id = publishers.publisher_id
WHERE
    books.num_pages > 50
GROUP BY
    publisher
ORDER BY
    cnt DESC;
''',
con = engine)

publisher  cnt
0                                        Penguin Books   42
1                                              Vintage   31
2                             Grand Central Publishing   25
3                                     Penguin Classics   24
4                                     Ballantine Books   19
..                                                 ...  ...
329                                         Turtleback    1
330  Atheneum Books for Young Readers: Richard Jack...    1
331                                     Penguin Signet    1
332                                    Victor Gollancz    1
333                      Harvard Business Review Press    1

[334 rows x 2 columns]

**Findings**:

- There are 334 publishers that has released the greatest number of books with more than 50 pages.


### Identify the author with the highest average book rating (look only at books with at least 50 ratings).


In [17]:
# Find the author with the highest average book rating

pd.io.sql.read_sql(
'''
SELECT
    authors.author_id,
    authors.author,
    COUNT(ratings.rating) AS cnt_rating,
    ROUND(AVG(ratings.rating), 2) AS avg_rating
FROM 
    books
    LEFT JOIN authors
        ON books.author_id = authors.author_id
    LEFT JOIN ratings
        ON books.book_id = ratings.book_id
GROUP BY
    authors.author_id,
    authors.author
HAVING
    COUNT(ratings.rating) >= 50
ORDER BY
    avg_rating DESC;
''',
con = engine)

author_id                                             author  cnt_rating  \
0         130                                     Diana Gabaldon          50   
1         236                         J.K. Rowling/Mary GrandPré         312   
2           3                                    Agatha Christie          53   
3         402                  Markus Zusak/Cao Xuân Việt Khương          53   
4         240                                     J.R.R. Tolkien         166   
5         499                           Roald Dahl/Quentin Blake          62   
6         376                                  Louisa May Alcott          54   
7         498                                       Rick Riordan          84   
8          39                                      Arthur Golden          56   
9         542                                       Stephen King         106   
10        302                                       John Grisham          70   
11        621                                    William Golding          71   
12        536                                    Sophie Kinsella          57   
13        291                                       Jodi Picoult          76   
14        445                                    Nicholas Sparks         111   
15        257                                    James Patterson          64   
16        235                                      J.D. Salinger          98   
17        630  William Shakespeare/Paul Werstine/Barbara A. M...          66   
18        469           Paulo Coelho/Alan R. Clarke/Özdemir İnce          57   
19        372                                         Lois Lowry          65   
20        106                                          Dan Brown         143   
21        195            George Orwell/Boris Grabnar/Peter Škerl          74   
22        554                                    Stephenie Meyer         160   
23        311                                     John Steinbeck          73   

    avg_rating  
0         4.30  
1         4.29  
2         4.28  
3         4.26  
4         4.24  
5         4.21  
6         4.20  
7         4.13  
8         4.11  
9         4.01  
10        3.97  
11        3.90  
12        3.88  
13        3.88  
14        3.88  
15        3.86  
16        3.85  
17        3.79  
18        3.79  
19        3.74  
20        3.74  
21        3.73  
22        3.66  
23        3.64

**Findings**:

- There are 18 authors with the highest average book rating (look only at books with at least 50 ratings).

### Find the average number of text reviews among users who rated more than 50 books.

Step by step:
- Using CTE (Common Table Expression), we created two new tables, as follows:
    1. First table to get the number of users who rated more than 50 books.
    2. Second table to get the number of users who review books.
- These two tables merge using INNER JOIN method on username.
- Get the average number of text reviews.

In [18]:
# Find the average number of text reviews among users who rated more than 50 books

pd.io.sql.read_sql(
'''
WITH user_more_than_50_ratings AS (
    SELECT
        username,
        COUNT(username) AS cnt_rating
    FROM 
        ratings
    GROUP BY
        username
    HAVING
        COUNT(username) > 50
)
, main AS (
    SELECT
        reviews.username,
        COUNT(reviews.review_id) AS cnt_reviews
    FROM
        reviews
        INNER JOIN user_more_than_50_ratings ON reviews.username = user_more_than_50_ratings.username
    GROUP BY
        reviews.username
    ORDER BY
        cnt_reviews DESC
)
SELECT
    ROUND(AVG(cnt_reviews), 0) AS avg_reviews
FROM
    main
''',
con = engine)

avg_reviews
0         24.0

### The EDA Conslusions

- The total of books released after January 1, 2000 are 819 books.
- There are 994 books that receive reviews from the users, with the range of reviews per book being 1 of 7 reviews and the range of average ratings is from a scale of 1.5 to 5.
- There are 334 publishers that has released the greatest number of books with more than 50 pages.
- There are 18 authors with the highest average book rating (look only at books with at least 50 ratings).
- The average number of text reviews among users who rated more than 50 books is 24.

## Overal Conslusions and Recommendations

**In the Data Pra-processing:**
- We have tables, as follows:
    1. The `books` table that contains data on books such as `book_id`, `author_id`, `title`, `num_pages`, `publication_date` and `publisher_id`.
    2. The `authors` table that contains data on authors such as `author_id` and `author name`.
    3. The `publishers` table that contains data on publishers such as `publisher_id` and `publisher name`.
    4. The `ratings` table that contains data on user ratings such as `rating_id` - refers to the rating order by the user, `book_id`, `username` — the name of the user who rated the book and `rating` - scale 1 to 5.
    5. The `reviews` table that contains data on customer reviews such as `review_id` - refers to the text review order by the user, `book_id`, `username` and `text` — the text of the review.
- There is no missing values.

**In the Exploratory Data Analysis (EDA) Section:**
- The total of books released after January 1, 2000 are 819 books.
- There are 994 books that receive reviews from the users, with the range of reviews per book being 1 of 7 reviews and the range of average ratings is from a scale of 1.5 to 5.
- There are 334 publishers that has released the greatest number of books with more than 50 pages.
- There are 18 authors with the highest average book rating (look only at books with at least 50 ratings).
- The average number of text reviews among users who rated more than 50 books is 24 reviews.

**Recommendations:**
- Choose the authors with the highest average ratings then it is possible to increase the sale of the next book.
- As we know that the average number of text reviews among users who rated more than 50 books in 24 reviews which is a quite big number, so it's a good insight to get deeper into the genre they read because the chance they give a text review is high.